Feature/add notes

ltwheeler opened the pull request:
Adds a way to store and view notes over time on both a course and instructor.

Issue:
https://trello.com/c/9q98Keux/1779-courses-add-a-way-to-store-and-view-notes-about-a-course-or-instructor-over-time

<img width="1240" alt="screen shot 2018-07-31 at 7 48 53 am" src="https://user-images.githubusercontent.com/1776859/43467473-99dfeffe-9496-11e8-9e22-b7aa2f10b4c4.png">
<img width="1249" alt="screen shot 2018-07-31 at 7 48 41 am" src="https://user-images.githubusercontent.com/1776859/43467478-9bfcd342-9496-11e8-8aad-83f044ce61ec.png">

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/entities/InstructorNote.java:
```diff
@@ -0,0 +1,98 @@
+@JsonDeserialize(using = InstructorNoteDeserializer.class)
+@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
+public class InstructorNote extends BaseEntity {
+  private long id;
```
Is this indenting inconsistent or are these updated linter rules?

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/entities/InstructorNote.java:
```diff
@@ -0,0 +1,98 @@
+
+  @JsonProperty("instructorId")
+  @Transient
+  public long getInstructorIdentification() {
```
Why is this getter and the one below it called "Identification" instead of Id?

ltwheeler replied:
We started following this pattern quite a while back, originally we did use getInstructorId() on transient methods like this one,
but I vaguely recall that after much debugging we found that names like getInstructorId() caused Hibernate to incorrectly infer things about the entity and it caused problems during repository actions. A slight change to the name so it no longer matched hibernate syntax but still made sense to humans seemed straightforward.

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/entities/InstructorNote.java:
```diff
@@ -0,0 +1,98 @@
+@SuppressWarnings("serial")
+@Entity
+@Table(name = "InstructorNotes")
+@JsonDeserialize(using = InstructorNoteDeserializer.class)
```
Do we need custom deserializers for plain objects?

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/entities/InstructorNote.java:
```diff
@@ -0,0 +1,98 @@
+    if(instructor != null) {
+      return instructor.getId();
+    } else {
+      return 0;
```
Why not return null with type Long instead of 0 and 'long'? '0' could technically be a valid ID. Same on line 95.

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaInstructorNoteService.java:
```diff
@@ -0,0 +1,64 @@
+    InstructorNote originalInstructorNote = this.findById(newInstructorNote.getId());
+
+    if(originalInstructorNote == null) {
+      return null;
```
What about throwing an exception instead of a silent null?

cthielen commented:
Tests are failing. It looks like you added a table without updating src/test/resources/schema.sql

ltwheeler commented:
- Fixed private staff comment modal.
- Updated styles to be taller and use all available space.
- Tweaked border style and added a dividing line.
<img width="241" alt="screen shot 2018-08-10 at 7 44 32 am" src="https://user-images.githubusercontent.com/1776859/43964101-f643d3c8-9c70-11e8-917b-043213e79eb6.png">

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/api/entities/InstructorNoteController.java:
```diff
@@ -0,0 +1,57 @@
+	@RequestMapping(value = "/api/schedules/{scheduleId}/instructors/{instructorId}/instructorNotes", method = RequestMethod.PUT, produces="application/json")
+	@ResponseBody
+	public InstructorNote createOrUpdateInstructorNote(@PathVariable long scheduleId,
+													   @PathVariable long instructorId,
```
Line up arguments

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/entities/InstructorNote.java:
```diff
@@ -0,0 +1,95 @@
+
+	@JsonProperty("instructorId")
+	@Transient
+	public long getInstructorIdentification() {
```
Should return Long, not long, if this is nullable.

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/entities/InstructorNote.java:
```diff
@@ -0,0 +1,95 @@
+
+	@JsonProperty("scheduleId")
+	@Transient
+	public long getScheduleIdentification() {
```
Should return Long, not long, if this is nullable.

cthielen commented on src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaInstructorNoteService.java:
```diff
@@ -0,0 +1,60 @@
+
+  private InstructorNote create(long scheduleId, long instructorId) {
+    InstructorNote instructorNote = new InstructorNote();
+    Schedule schedule = scheduleRepository.findOne(scheduleId);
```
No need for the temporary variables 'schedule' and 'instructor' if you never check them and only use them once.
